asserts/asserts.go

```diff
@@ -257,10 +277,114 @@ func buildAssertion(headers map[string]string, body, content, signature []byte)
 	return assert, nil
 }
 
+func writeHeader(buf *bytes.Buffer, headers map[string]string, name string) {
+	buf.WriteString(name)
+	buf.WriteString(": ")
+	buf.WriteString(headers[name])
+	buf.WriteByte('\n')
+}
+
+var marker = struct{}{}
+
+// TODO: expose this in some form on Database appropriately
+func buildAndSign(assertType AssertionType, headers map[string]string, body []byte, privKey *packet.PrivateKey) (Assertion, error) {
+	finalHeaders := make(map[string]string, len(headers))
+	for name, value := range headers {
+		finalHeaders[name] = value
+	}
+	bodyLength := len(body)
+	finalBody := make([]byte, bodyLength)
+	copy(finalBody, body)
+	finalHeaders["type"] = string(assertType)
+	finalHeaders["body-length"] = strconv.Itoa(bodyLength)
+
+	if _, err := checkMandatory(finalHeaders, "authority-id"); err != nil {
+		return nil, err
+	}
+
+	reg, err := checkAssertType(assertType)
+	if err != nil {
+		return nil, err
+	}
+
+	revision, err := checkRevision(finalHeaders)
+	if err != nil {
+		return nil, err
+	}
+
+	buf := new(bytes.Buffer)
+	writeHeader(buf, finalHeaders, "type")
+	writeHeader(buf, finalHeaders, "authority-id")
+	if revision > 0 {
+		writeHeader(buf, finalHeaders, "revision")
+	} else {
+		delete(finalHeaders, "revision")
+	}
+	written := map[string]struct{}{
```

Comment:
I wouldn't bother saving the 1 byte of bool. It's much more convenient to have this as map[string]bool. It even allows you to put the its false zero value to good use, by doing `if !written[name] {` below, which looks very nice.